java/arcs/core/util/UnionFind.kt

```diff
@@ -13,18 +13,30 @@ package arcs.core.util
 
 /**
  * A union-find datastructure for computing equivalence classes consisting of [E] instances.
- * This datastructure is not thread safe.
+ * Optionally, a value of type [I] may be associated with the equivalence class. If the client
+ * does not need to associated additional information with the class, they may use the [Unit]
+ * type for [I]. This datastructure is not thread safe.
  */
-class UnionFind<E> {
+class UnionFind<E, I> {
     /**
-     * Unifies the equivalence classes of elements [e1] and [e2]. If either
-     * element is not present in any set, it is added.
+     * Unifies the equivalence classes of elements [e1] and [e2]. If either element is not present
+     * in any set, it is added. The [combine] function determines how the information associated
+     * with the two equivalence classes are merged. The default [combine] method returns the
+     * information associated with the new root.
      */
-    fun union(e1: E, e2: E) {
+    fun union(
+        e1: E,
+        e2: E,
+        combine: (destInfo: I?, srcInfo: I?) -> I? = {
+            destInfo: I?, _: I? -> destInfo
```

Comment:
Lambda params go on the same line as the opening brace.

```kotlin
combine: (destInfo: I?, srcInfo: I?) -> I? = { destInfo, _ ->
     destInfo
}
```

That said, this is also short enough to all be on one line.